lib/net.c

```diff
@@ -74,35 +74,40 @@ ServerPtr server_open(DispatchPtr dispatch, const char *listenaddr, unsigned int
 		log_debug("Opening connection on %s:%u", listenaddr, port);
 		server->listener = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
 		if (server->listener != -1) {
-			struct sockaddr_in all_if;
-			memset(&all_if, 0, sizeof(all_if));
-			all_if.sin_family = AF_INET;
-			all_if.sin_port = htons((uint16_t) port);
-			if (inet_pton(AF_INET, listenaddr, &all_if.sin_addr) == 1) {
-				if (bind(server->listener, (struct sockaddr *) &all_if, sizeof(all_if)) == 0) {
-					if (listen(server->listener, MAX_CONNECTIONS) == 0) {
-						server->dispatch = dispatch;
-						if (dispatch) {
-							log_debug("Registering server socket %d", server->listener);
-							dispatch_add(dispatch, server->listener, POLLIN, server_listener_ready, server_listener_error, NULL, server);
+			if (setsockopt(server->listener, SOL_SOCKET, SO_REUSEADDR, & (int) {1}, sizeof(int)) == 0) {
```

Comment:
Let's see here: `& (int) { 1 }`
You're creating an array literal of `int` with one member. The equivalent type of this literal is `int*`. You then typecast this `int*` to `int`, effectively converting a pointer to integer of possibly incompatible size. Then you construct a pointer to this `int*` by referencing it with `&`.
That reeks of type confusion to me - even if it may work.
Please move this array literal in front of the setsockaddr and reference it properly so it's clearer to understand and the risk of type confusion is eliminated.

Reply:
I've had a look at this again. This is not a type confusion. The initializer `(int) {1}` does not create an int array, but a single int, whose reference is then passed to setsockopt. So it isn't wrong, just ugly :)

Reply:
Weird.
Still, hard to understand code is a bug risk to bug creep, and spelling out the variable incurs no runtime impact.